### PR TITLE
feat: notify controllers when outputs are ready to be destroyed

### LINKED ISSUE #34: notify controller on outputs being ready for destroy
When controller output enters teardDown phase and has no finalizers, controller should be notified to run reconcile loop

### PR DESCRIPTION
**Pull request by smira**
Controllers should use two-phased approach when destroying their
outputs:

* first, resource is marked as being torn down (`Teardown()`), this also
prevents adding new finalizers
* second, when finalizers are removed by dependent controllers,
controller can finally completely remove its out with `Destroy()`

This problem is that controller should wait for the finalizers to be
empty between phases one and two. Waiting is possible, but controller
might not want to block other reconcilation activities while waiting for
the output to be ready to be destroyed.

This PR addresses this issue by allowing to declare controller output as
an input of special kind `InputDestroyReady` which means controller is
only interested to be notified when output is ready to be destroyed.

Fixes #34

Signed-off-by: Andrey Smirnov <smirnov.andrey@gmail.com>

### CODE REVIEW

File: api/v1alpha1/runtime.proto [hunk 1]
```diff
@@ -127,6 +127,16 @@ enum ControllerInputKind {
...
+    // DestroyReady inputs provide notifications to the controllers when some resource
+    // is ready to be destroyed (all finalizers are removed).
+
+    // Usually outputs are declared as DestroyReady inputs to trigger reconcile loop
+    // when resource is ready to be destroy.
+    //
+    // Controller will be notified when the resource enters "teardown" phase and has no
+    // finalizers attached.
+    // Resources are filtered to be owned by the controller.
+    DESTROY_READY = 2;
```
Comment (andrewrynhard) [inline api/v1alpha1/runtime.proto:139]: This feels a little out of place here, but I am not comfortable enough with the code yet to suggest anything different.

Comment (smira) [reply api/v1alpha1/runtime.proto:139]: I was going back and forth on the design here, as it certainly feels like an output property, but outputs don't have namespaces, so I got stuck as a watch implies having a namespace.

Comment (smira) [reply api/v1alpha1/runtime.proto:139]: probably we can come up with some other concept here to express it better

File: api/v1alpha1/runtime.proto [hunk 2]
```diff
@@ -127,6 +127,16 @@ enum ControllerInputKind {
     //
     // Controller can add finalizers on strong inputs.
     STRONG = 1;
+    // DestroyReady inputs provide notifications to the controllers when some resource
+    // is ready to be destroyed (all finalizers are removed).
+
+    // Usually outputs are declared as DestroyReady inputs to trigger reconcile loop
+    // when resource is ready to be destroy.
```
Comment (Unix4ever) [inline api/v1alpha1/runtime.proto:134]:
```suggestion
    // when resource is ready to be destroyed.
```

**Review by Unix4ever — APPROVED** [review]

File: api/v1alpha1/runtime.proto (hunk 1 above)
Comment (andrewrynhard) [reply api/v1alpha1/runtime.proto:139]: Where did we land with this?

Comment (smira) [reply api/v1alpha1/runtime.proto:139]: I don't have any better idea than what we have here, it's more about the API, less about the actual implementation.

File: api/v1alpha1/runtime.proto (hunk 2 above)
Comment (smira) [reply api/v1alpha1/runtime.proto]: thanks, fixed!

**Review by Unix4ever — APPROVED** [review]

**Pull request merged**